lib/linked_list.rb

```diff
@@ -12,152 +12,309 @@ def initialize(value, next_node = nil)
-    # Time Complexity:  
-    # Space Complexity
-    def add_first(value)
-      raise NotImplementedError
+  # method to find if the linked list contains a node with specified value
+  # returns true if found, false otherwise
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def search(value)
+    current = @head
+
+    until current.next == nil
+      return true if value == current.data
+
+      current = current.next
     end
 
-    # method to find if the linked list contains a node with specified value
-    # returns true if found, false otherwise
-    # Time Complexity:  
-    # Space Complexity
-    def search(value)
-      raise NotImplementedError
+    return false
+  end
+
+  # method to return the max value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def find_max
+    return nil if !@head
+
+    current = @head.next
+    max_value = @head.data
+
+    until !current
+      max_value = current.data if current.data > max_value
+      current = current.next
     end
 
-    # method to return the max value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_max
-      raise NotImplementedError
+    return max_value
+  end
+
+  # method to return the min value in the linked list
+  # returns the data value and not the node
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def find_min
+    return nil if !@head
+
+    current = @head.next
+    min_value = @head.data
+
+    until !current
+      min_value = current.data if current.data < min_value
+      current = current.next
     end
 
-    # method to return the min value in the linked list
-    # returns the data value and not the node
-    # Time Complexity:  
-    # Space Complexity
-    def find_min
-      raise NotImplementedError
+    return min_value
+  end
+
+  # method that returns the length of the singly linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def length
+    return 0 if !@head
+
+    count = 1
+    current = @head.next
+
+    until !current
+      count += 1
+      current = current.next
     end
 
+    return count
+  end
 
-    # method that returns the length of the singly linked list
-    # Time Complexity:  
-    # Space Complexity
-    def length
-      raise NotImplementedError
+  # method that returns the value at a given index in the linked list
+  # index count starts at 0
+  # returns nil if there are fewer nodes in the linked list than the index value
+  # Time Complexity: O(n)
+  # Space Complexity: O(1)
+  def get_at_index(index)
+    count = 0
+    current = @head
+    while current != nil
+      if count == index
+        return current.data
+      end
+      current = current.next
+      count += 1
     end
+    return nil
+  end
+
+  # method to print all the values in the linked list
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
+  def visit

```

Comment:
This method isn't printing values in the list.  Also the space complexity isn't O(n)